File: scripts/Random Flip Matrix/Random Flip Matrix.py
from random import randint

class Solution:

    def __init__(self, m: int, n: int):
        self.m = m
        self.n = n
        self.ones = set()

    def flip(self) -> List[int]:
        i, j = randint(0, self.m - 1), randint(0, self.n - 1)
        while (i, j) in self.ones:
            i, j = randint(0, self.m - 1), randint(0, self.n - 1)
        self.ones.add((i, j))
        return [i, j]

    def reset(self) -> None:
        self.ones.clear()

File: scripts/algorithms/R/Random Flip Matrix/Random Flip Matrix.js
// Runtime: 80 ms (Top 83.33%) | Memory: 61.30 MB (Top 16.67%)

var Solution = function(m, n) {
   this.length = m * n;
   this.range = new Array(this.length);
   this.n = n;
   this.pointer = 0;    
};

Solution.prototype.flip = function() {
    let limit = this.length - this.pointer - 1;
    let index = ~~(Math.random() * limit) + this.pointer;

    if(this.range[index] === undefined) this.range[index] = index;
    if(this.range[this.pointer] === undefined) this.range[this.pointer] = this.pointer;

    let tmp = this.range[index];
    this.range[index] = this.range[this.pointer];
    this.range[this.pointer] = tmp;

    let a = ~~(this.range[this.pointer] / this.n);
    let b = this.range[this.pointer] % this.n;

    this.pointer++;
    return[a, b];
};

Solution.prototype.reset = function() {
    this.pointer = 0;
};

